### DISCUSSION TITLE
Extend Python classes

### DISCUSSION THREAD

**Discussion by gents83** [PyO3/pyo3#1989] [Questions]
Hi!
This crate is great and the more I use it, the more I discover!

I saw that we have subclass and extends for pyclass in the same Rust file/module, but is there any way to extend a class declared and implemented in a Python module from Rust?

**Comment by birkenfeld** [comment]
Nice to hear you like PyO3! The big question here is what you want to do with your class.

If you want to access it from the Rust side, the Rust code can't know anything about the API coming from Python, so you're limited to use the generic Python access API (`.call()`, `.get_attr()` and so forth), which is not a nice experience.

If you want to use it from Python, then you might be able to do a trick: create the Rust class as a mixin, and make a new class (you can do that in Python or in Rust) which inherits from the Rust mixin and the Python base class (in that order).

However, if you overwrite methods and want to call its base methods, it will get complicated. I don't think we have an API that mirrors `super()` in Python, but that may be possible to add.

**Reply by gents83** [reply]
My actual idea was to try to create nodes for a nodetree of Python directly in Rust.
Ok so probably what I did was already the only way to do that :)

To be super clear, what I did was actually having a basic trait in Rust, creating struct declared as pyclass (with various subclass, extends, etc) in Rust and then going back in Python I created new types using type() and specifying the different extends.

It works but it's not super "polished" due to the double bounce Python<->Rust<->Python

**Comment by davidhewitt** [comment]
Are you looking for #991? It should be possible eventually, however we're unfortunately not there yet (and it hasn't been the highest-priority item for me).

**Reply by gents83** [reply]
Yep exactly 😁 
Tnx for pointing out to the whole discussions about it!
